File: examples/06_properties.cpp
// 06 - Custom properties and status bits (C++).
//   Attach named typed data to any element kind; toggle status flags.
#include "simplemesh/Mesh.h"
#include <iostream>

int main() {
    sm::Mesh m;
    auto a = m.add_vertex({0, 0, 0});
    auto b = m.add_vertex({1, 0, 0});
    auto c = m.add_vertex({0, 1, 0});
    auto f = m.add_triangle(a, b, c);

    // a per-vertex scalar "quality"
    auto quality = m.add_vertex_property<double>("quality", 0.0);
    m.property(quality, a) = 0.9;
    m.property(quality, b) = 0.5;
    std::cout << "quality[a] = " << m.property(quality, a) << "\n";
    std::cout << "n vertex properties: " << m.n_vertex_properties() << "\n";

    // a per-face Vec3 "color"
    auto color = m.add_face_property<sm::Vec3>("color", {0, 0, 0});
    m.property(color, f) = {1, 0, 0};
    std::cout << "color[f]   = " << m.property(color, f) << "\n";

    // status bits
    m.set_selected(a);
    m.set_locked(b);
    std::cout << "a selected : " << m.is_selected(a) << "\n";
    std::cout << "b locked   : " << m.is_locked(b) << "\n";
    std::cout << "a status word: " << m.get_status(a)
              << " (SELECTED = " << (int)sm::Mesh::SELECTED << ")\n";
    return 0;
}

File: examples/06_properties.py
"""06 - Custom properties and status bits.

Attach named typed data (float / int / Vec3) to any element kind, and toggle the
built-in status flags. Both grow with the mesh automatically.
Run:  PYTHONPATH=build/bin python examples/06_properties.py
"""
import simplemesh as sm

m = sm.Mesh()
a = m.add_vertex(sm.Vec3(0, 0, 0))
b = m.add_vertex(sm.Vec3(1, 0, 0))
c = m.add_vertex(sm.Vec3(0, 1, 0))
f = m.add_triangle(a, b, c)

# a per-vertex scalar "quality"
quality = m.add_vertex_float_property("quality", 0.0)
m.set_vertex_float(quality, a, 0.9)
m.set_vertex_float(quality, b, 0.5)
print("quality[a] =", m.get_vertex_float(quality, a))
print("n vertex properties:", m.n_vertex_properties)

# a per-face Vec3 "color"
color = m.add_face_vec3_property("color", sm.Vec3(0, 0, 0))
m.set_face_vec3(color, f, sm.Vec3(1, 0, 0))
print("color[f]   =", m.get_face_vec3(color, f))

# status bits
m.set_vertex_selected(a, True)
m.set_vertex_locked(b, True)
print("a selected :", m.is_vertex_selected(a))
print("b locked   :", m.is_vertex_locked(b))
print("a status word:", m.vertex_status(a), "(SELECTED =", sm.SELECTED, ")")
